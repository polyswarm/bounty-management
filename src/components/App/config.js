const config = {
  url: process.env.REACT_APP_HOST,
};
export default config;
