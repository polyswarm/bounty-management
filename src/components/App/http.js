import BigNumber from 'bignumber.js';
class HttpApp {
  constructor(url, ws) {
    this.url = url;
    this.ws = ws;
  }

  getUnlockedWallet() {
    return fetch(this.url+'/accounts/active')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => !json.locked)
      .catch(() => false);
  }

  getWallets() {
    return fetch(this.url+'/accounts')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .catch(() => []);
  }

  getBounty(bounty) {
    return fetch(this.url+'/bounties/'+bounty.guid)
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw Error('Cannot get bounties.');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(bounty => this.getAssertionsForBounty(bounty))
      .then(bountyAssertions => this.getArtifactsForBounty(bountyAssertions))
      .then(bounty => {
        const assertions = bounty.assertions;
        let files = bounty.artifacts;
        assertions.forEach((assertion) => {
          assertion.verdicts.forEach((verdict, index) => {
            const file = files[index];
            if (!verdict) {
              file.good++;
            }
            file.total++;
            file.assertions.push({
              author: assertion.author,
              bid: assertion.bid,
              verdict: verdict,
              metadata: assertion.metadata
            });
            files[index] = file;
          });
        });
        bounty.artifacts = files;
        return bounty;
      })
      .catch(() => null);
  }

  getArtifactsForBounty(bounty) {
    return fetch(this.url+'/artifacts/'+bounty.uri)
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(files => {
        return files.map((file) => {
          return file.name;
        });
      })
      .then(filesnames => {
        return filesnames.map((name) => {
          const trimmed = name.trim();
          return {name: trimmed, good: 0, total: 0, assertions: []};
        });
      })
      .then(files => {
        bounty.artifacts = files;
        return bounty;
      });
  }

  getAssertionsForBounty(bounty) {
    return fetch(this.url+'/bounties/'+bounty.guid+'/assertions')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(assertions => {
        return assertions.map((assertion) => {
          const bid = new BigNumber(assertion.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
          return {
            author: assertion.author,
            bid: bid,
            verdicts: assertion.verdicts,
            metadata: assertion.metadata,
          };
        });
      })
      .then(filtered => {
        bounty.assertions = filtered;
        return bounty;
      });
  }

  listenForAssertions(assertionAddedCallback) {
    // attach to websocket
    // anytime we get an assertion, check if it matches a guid
    // if it does, add it to the assertions for that object
    const ws = this.ws;
    const websocket = new WebSocket(ws);

    websocket.onmessage = (event) => {
      const message = JSON.parse(event.data);

      if (message.event === 'assertion') {
        const body = message.data;
        const bid = new BigNumber(body.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
        const assertion = {
          guid: body.bounty_guid,
          bid: bid,
          verdicts: body.verdicts,
          metadata: body.metadata,
          author: body.author,
        };
        assertionAddedCallback(assertion);
      }
    };
  }
}
export default HttpApp;
