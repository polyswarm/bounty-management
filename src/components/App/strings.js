const strings = {
  create: 'Create a Bounty',
  bountyError: 'Error: Failed to Post Bounty',
  requestAllData: 'Refreshing bounties',
  requestWallets: 'Updating Wallets',
  requestGetBounty: 'Getting Bounty'
};

export default strings;
