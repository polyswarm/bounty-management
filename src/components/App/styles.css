html,
body,
#root {
  height: 100%;
}

body {
  font-family: 'Roboto', sans-serif;
  margin: 0;
  overflow: hidden;
  padding: 0;
}

.BountyManager {
  color: white;
  height: 100%;
  left: 0;
  position: absolute;
  right: 0;
  text-align: center;
  top: 0;
}

.hex-background {
  background-color: #8500ff;
  background-image: url('/img/hexgrid_2x.png');
  background-size: 129px 149px;
}

.BountyManager-Intro {
  font-size: large;
}
