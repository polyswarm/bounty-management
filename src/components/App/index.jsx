// Vendor imports
import React, { Component } from 'react';
// Bounty imports
import BountyCreate from '../BountyCreate';
import BountyInfo from '../BountyInfo';
import Sidebar from '../Sidebar';
import Header from '../Header';
import Welcome from '../Welcome';
// Component imports
import HttpApp from './http';
import config from './config';
import strings from './strings';
import './styles.css';

class App extends Component {
  constructor(props) {
    super(props);
    const {bounties, first} = this.preloadLocalStorage();
    this.http = new HttpApp(config.url, config.wsUrl);
    this.state = {
      isUnlocked: false,
      walletList: [],
      active: 0,
      bounties: bounties,
      create: false,
      first: first,
    };

    this.onAddBounty = this.onAddBounty.bind(this);
    this.onRemoveBounty = this.onRemoveBounty.bind(this);
    this.onSelectBounty = this.onSelectBounty.bind(this);
    this.onCreateBounty = this.onCreateBounty.bind(this);
    this.onCloseWelcome = this.onCloseWelcome.bind(this);
    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
    this.getData = this.getData.bind(this);
    this.getWallets = this.getWallets.bind(this);
  }

  componentDidUpdate(_, prevState) {
    const { state: { bounties } } = this;
    const { bounties: prevBounties } = prevState;
    const storageOutOfSync =  JSON.stringify(bounties) !== JSON.stringify(prevBounties);
    if(storageOutOfSync) {
      this.storeBounties(bounties);
    }
  }

  componentDidMount() {
    this.getWallets();
    this.getData();
  }

  render() {
    const {url} = config;
    const { state: { active, bounties, create, first, isUnlocked, walletList } } = this;

    return (
      <div className='App'>
        {first && (
          <Welcome onClick={this.onCloseWelcome}/>
        )}
        {!first && (
          <React.Fragment>
            <Sidebar bounties={bounties}
              active={active}
              remove={this.onRemoveBounty}
              select={this.onSelectBounty}/>
            <Header title={(bounties.length === 0 || create || active < 0) ? strings.create : bounties[active].guid}
              create={create}
              onClick={this.onCreateBounty}/>
            <div className='App-Content'>
              { (bounties.length === 0 || create || active < 0 ) && (
                <BountyCreate url={url}
                  isUnlocked={isUnlocked}
                  walletList={walletList}
                  onWalletChange={this.onWalletChangeHandler}
                  addBounty={this.onAddBounty}/>
              )}
              { !create && active >=0 && active < bounties.length && (
                <BountyInfo bounty={bounties[active]}/>
              )}
            </div>
          </React.Fragment>
        )}
      </div>
    );
  }

  onAddBounty(guid) {
    const bounties = this.state.bounties.slice();
    const bounty = {
      guid: guid,
      update: false,
      author: '',
      amount: '',
      artifactURI: '',
      expirationBlock: '',
      resolved: '',
      verdicts: '',
    };
    bounties.push(bounty);
    this.setState({bounties: bounties});
  }

  onCreateBounty() {
    this.setState({create: true, active: -1});
  }

  onCloseWelcome() {
    this.setState({first: false});
    this.markSeen();
  }

  onRemoveBounty(index) {
    const bounties = this.state.bounties.slice();
    if (index !== null && index >= 0 && index < bounties.length) {
      bounties.splice(index, 1);
      this.setState({bounties: bounties});
    }
  }

  onSelectBounty(index) {
    const bounties = this.state.bounties.slice();
    if (index !== null && index >= 0 && index < bounties.length) {
      const bounty = bounties[index];
      bounty.updated = false;
      bounties[index] = bounty;
      this.setState({active: index, create: false, bounties: bounties});
    }
  }

  onWalletChangeHandler(store) {
    this.setState({isUnlocked: store});
  }

  pullLocalBounties() {
    const bounties = this.state.bounties.slice();
    return bounties;
  };

  updateOnAssertion(assertion) {
    const bounties = this.state.bounties.slice();
    const guid = assertion.guid;
    const index = bounties.findIndex((bounty) => bounty.guid === guid);
    // let index = -1;
    // bounties.forEach((bounty, i) => {
    //   if (bounty.guid === guid) {
    //     index = i;
    //   }
    // });
    if (index >= 0) {
      const bounty = bounties[index];
      const a = {};
      a[assertion.author] = {
        bid: assertion.bid,
        metadata: assertion.metadata,
        verdicts: assertion.verdicts,
      };
      bounty.assertsions.push(a);
      const modified = bounty.artifacts.map((file, index) => {
        const f = file;
        if (!a.verdicts[index]) {
          f.good++;
        }
        f.total++;
        file.assertions.push({
          author: assertion.author,
          bid: assertion.bid,
          verdict: a.verdicts[index],
          metadata: assertion.metadata
        });
        return f;
      });
      bounty.artifacts = modified;
      bounty.update = true;
      bounties[index] = bounty;
    }
    this.setState({bounties: bounties});
  }

  getData() {
    const http = this.http;
    http.listenForAssertions(this.pullLocalBounties, this.updateOnAssertion);
    const bounties = this.state.bounties.slice();
    const promises = bounties.map((bounty) => {
      return http.getBounty(bounty)
        .then(b => {
          if (Object.keys(b) !== Object.keys(bounty)) {
            b.update = true;
          }
          return b;
        })
        .catch(() => bounty);
    });
    Promise.all(promises).then((values) => {
      // get updated state after download finishes
      const bounties = this.state.bounties.slice();
      values.forEach((value) => {
        const foundIndex = bounties.findIndex((bounty) => bounty.guid === value.guid);
        if (foundIndex >= 0) {
          bounties[foundIndex] = value;
        }
      });
      this.setState({bounties: bounties});
    });
  }

  getWallets() {
    const http = this.http;
    http.getWallets()
      .then(accounts => {
        this.setState({walletList: accounts});
      });

    http.getUnlockedWallet()
      .then((success) => this.setState({isUnlocked: success}));
  }

  storeBounties(bounties) {
    if (this.hasLocalStorage()) {
      localStorage.setItem('bounties', JSON.stringify(bounties));
    }
  }

  hasLocalStorage() {
    try {
      localStorage.setItem('x', 'y');
      localStorage.removeItem('x');
      return true;
    } catch(e) {
      return false;
    }
  }

  markSeen() {
    if (this.hasLocalStorage()) {
      localStorage.setItem('seen', JSON.stringify(true));
    }
  }

  preloadLocalStorage() {
    if (this.hasLocalStorage) {
      const bounties = JSON.parse(localStorage.getItem('bounties')) || [];
      const first = !JSON.parse(localStorage.getItem('seen'));
      return {bounties, first};
    } else {
      return {bounties: [], first: true};
    }
  }
}

export default App;
