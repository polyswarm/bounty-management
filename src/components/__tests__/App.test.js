import React from 'react';
import {render, mount} from 'enzyme';
import {renderToJson, mountToJson} from 'enzyme-to-json';
import LocalStorage from '../__mocks__/localstorage';
import App from '../App';

beforeEach(() => {
  localStorage.clear();
});

it('renders without crashing', () => {
  const wrapper = render(<App />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('shows create when no bounties found', () => {
  const wrapper = mount(<App />);
  wrapper.setState({first: false});
  expect(wrapper.find('.Bounty-Create')).toHaveLength(1);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('shows manager when at least one bounty & active selects it', () => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}];
  const active = 0;
  wrapper.setState({first: false, bounties: bounties, active: active});
  expect(wrapper.find('.Manager')).toHaveLength(1);
  expect(mountToJson(wrapper)).toMatchSnapshot();
});

it('shows manager when at least one bounty & active is negative', () => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}];
  const active = -1;
  wrapper.setState({first: false, bounties: bounties, active: active});
  expect(wrapper.find('.Manager')).toHaveLength(1);
  expect(mountToJson(wrapper)).toMatchSnapshot();
});

// it('shows welcome screen on first load', () => {
//   const wrapper = mount(<App />);
//   expect(wrapper.find('.Welcome')).toHaveLength(1);
// });

it('calls select when a sidebar item is clicked', () => {
  const select = spyOn(App.prototype, 'onSelectBounty');
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});
  wrapper.find('.item-0').simulate('click');
  expect(select).toHaveBeenCalledTimes(1);
  expect(select).toHaveBeenCalledWith(0);
});

it('updates the state when onSelectBounty called',() => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onSelectBounty(1);

  expect(setState).toHaveBeenCalledWith({active: 1});
});

it('does not update the state when onSelectBounty called with negative',() => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onSelectBounty(-1);

  expect(setState).toHaveBeenCalledTimes(0);
});

it('does not update the state when onSelectBounty called with null',() => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onSelectBounty(null);

  expect(setState).toHaveBeenCalledTimes(0);
});

it('does not update the state when onSelectBounty called with out of bounds',() => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onSelectBounty(1000);

  expect(setState).toHaveBeenCalledTimes(0);
});

it('calls remove when a sidebar item remove is clicked', () => {
  const remove = spyOn(App.prototype, 'onRemoveBounty');
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});
  wrapper.find('.item-0').simulate('mouseEnter');
  wrapper.find('.Remove-Button').simulate('click');
  expect(remove).toHaveBeenCalledTimes(1);
  expect(remove).toHaveBeenCalledWith(0);
});

it('updates the state when onRemoveBounty called',() => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onRemoveBounty(1);

  expect(setState).toHaveBeenCalledWith({bounties:[{guid:'asdf'}]});
});

it('removes the value at the index passed in onRemoveBounty', () => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onRemoveBounty(0);

  expect(setState).toHaveBeenCalledWith({bounties:[{guid:'demo'}]});
});

it('doesn\'t remove anything if onRemoveBounty called with negative', () => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onRemoveBounty(-1);

  expect(setState).toHaveBeenCalledTimes(0);
});

it('doesn\'t remove anything if onRemoveBounty called with null', () => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onRemoveBounty(null);

  expect(setState).toHaveBeenCalledTimes(0);
});

it('doesn\'t remove anything if onRemoveBounty called with out of bounds', () => {
  const wrapper = mount(<App />);
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const active = 0;
  wrapper.setState({bounties: bounties, active: active});

  const setState = spyOn(App.prototype, 'setState');
  const instance = wrapper.instance();
  instance.onRemoveBounty(2);

  expect(setState).toHaveBeenCalledTimes(0);
});

it('calls setState during onAddBounty', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();

  const setState = spyOn(App.prototype, 'setState');
  instance.onAddBounty('asdf');

  expect(setState).toHaveBeenCalledWith({bounties: [{
    guid: 'asdf',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});
});

it('calls setState during onAddBounty with existing values', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  wrapper.setState({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});

  const setState = spyOn(App.prototype, 'setState');

  instance.onAddBounty('asdf');

  expect(setState).toHaveBeenCalledWith({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  },
  {
    guid: 'asdf',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});
});

it('calls storeBounties after onAddBounty', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();

  const storeBounties = spyOn(App.prototype, 'storeBounties');
  instance.onAddBounty('asdf');

  expect(storeBounties).toHaveBeenCalledWith([{
    guid: 'asdf',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]);
});

it('calls setState during onRemoveBounty', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  wrapper.setState({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});

  const setState = spyOn(App.prototype, 'setState');
  instance.onRemoveBounty(0);

  expect(setState).toHaveBeenCalledWith({bounties:[]});
});

it('calls storeBounties after onRemoveBounty', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  wrapper.setState({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});

  const storeBounties = spyOn(App.prototype, 'storeBounties');
  instance.onRemoveBounty(0);

  expect(storeBounties).toHaveBeenCalledWith([]);
});

it('calls setState during onUpdateBounty', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  wrapper.setState({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});

  const setState = spyOn(App.prototype, 'setState');
  instance.onUpdateBounty('existing', '','','','','','');

  expect(setState).toHaveBeenCalledWith({bounties:[{
    guid: 'existing',
    update: true,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});
});

it('calls storeBounties after onUpdateBounty', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  wrapper.setState({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});

  const storeBounties = spyOn(App.prototype, 'storeBounties');
  instance.onUpdateBounty('existing', '','','','','','');

  expect(storeBounties).toHaveBeenCalledWith([{
    guid: 'existing',
    update: true,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]);
});

it('does not call storeBounties after onUpdateBounty with null guid', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  wrapper.setState({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});

  const storeBounties = spyOn(App.prototype, 'storeBounties');
  instance.onUpdateBounty(null, '','','','','','');

  expect(storeBounties).toHaveBeenCalledTimes(0);
});

it('does not call storeBounties after onUpdateBounty with guid that does not match', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  wrapper.setState({bounties: [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }]});

  const storeBounties = spyOn(App.prototype, 'storeBounties');
  instance.onUpdateBounty('asdf', '','','','','','');

  expect(storeBounties).toHaveBeenCalledTimes(0);
});

it('doesn\'t call storeBounties when setState called with identical set of bounties', () => {
  const wrapper = mount(<App />);
  const bounties = [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }];
  const storeBounties = spyOn(App.prototype, 'storeBounties');
  wrapper.setState({bounties: bounties});
  wrapper.setState({bounties: bounties});

  expect(storeBounties).toHaveBeenCalledTimes(1);
});

it('stores bounties into localstore when storeBounties is called', () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  const bounties = [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }];

  const setItem = jest.spyOn(LocalStorage.prototype, 'setItem');

  instance.storeBounties(bounties);

  expect(setItem).toHaveBeenCalledTimes(2);
  expect(setItem.mock.calls[1]).toEqual(['bounties', bounties]);
});

it('reads bounties from localStorage and puts into state on startup', () => {
  const bounties = [{
    guid: 'existing',
    update: false,
    author: '',
    amount: '',
    artifactURI: '',
    expirationBlock: '',
    resolved: '',
    verdicts: '',
  }];
  localStorage.setItem('bounties', bounties);

  const wrapper = mount(<App />);
  const instance = wrapper.instance();

  expect(instance.state.bounties).toBe(bounties);
});
