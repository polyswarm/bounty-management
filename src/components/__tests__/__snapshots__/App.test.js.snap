// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="App"
>
  <div
    class="Welcome"
  >
    <div
      class="Welcome-Jumbo"
    >
      <h1>
        Welcome to Polyswarm,
      </h1>
      <h2>
        The first 
        <em>
          decentralized
        </em>
         antivirus marketplace.
      </h2>
      <p>
        Simply upload files to create your own bounties and then let the microengines on Polyswarm analyze your file.
      </p>
      <button
        class="Button"
      >
        Get Started!
      </button>
    </div>
  </div>
</div>
`;

exports[`shows BountyInfo when at least one bounty & active is negative 1`] = `
<App>
  <div
    className="App"
  >
    <Sidebar
      active={-1}
      bounties={
        Array [
          Object {
            "guid": "asdf",
          },
        ]
      }
      remove={[Function]}
      select={[Function]}
    >
      <div
        className="Sidebar"
      >
        <header
          className="Sidebar-Header"
        >
          <img
            alt="Polyswarm logo"
            src="/img/polyswarm.svg"
          />
        </header>
        <ul
          className="Sidebar-Content"
        >
          <ListItem
            active={false}
            className="item-0"
            key="asdf"
            onClick={[Function]}
            remove={[Function]}
          >
            <li
              className="ListItem"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="ListItem-Child"
              >
                asdf
              </div>
            </li>
          </ListItem>
        </ul>
      </div>
    </Sidebar>
    <Header
      onClick={[Function]}
      title="Create a Bounty"
    >
      <header
        className="Header"
      >
        <h1>
          Create a Bounty
        </h1>
        <Button
          className="Header-Button"
          onClick={[Function]}
        >
          <button
            className="Button"
            onClick={[Function]}
          >
            + Bounty
          </button>
        </Button>
      </header>
    </Header>
    <div
      className="App-Content"
    >
      <BountyCreate
        addBounty={[Function]}
        url="https://localhost:8080"
      >
        <div
          className="Bounty-Create"
        >
          <div
            className="Container"
          >
            <DropTarget
              onFilesSelected={[Function]}
            >
              <div
                className="Drop-Target"
                onDragEnter={[Function]}
                onDragLeave={[Function]}
                onDragOver={[Function]}
                onDrop={[Function]}
              >
                <div>
                  <p>
                    Drag and drop your files here, or 
                  </p>
                  <FileButton
                    onFileSelected={[Function]}
                  >
                    <form>
                      <input
                        className="hidden"
                        id="file"
                        onChange={[Function]}
                        type="file"
                      />
                      <label
                        htmlFor="file"
                      >
                        click here.
                      </label>
                    </form>
                  </FileButton>
                </div>
              </div>
            </DropTarget>
            <FileList
              clear={[Function]}
              files={Array []}
              removeFile={[Function]}
            >
              <div
                className="File-List"
              >
                <header
                  className="File-List-Header"
                >
                  <p
                    className="File-List-Title"
                  >
                    0
                     
                    Files
                  </p>
                  <Button
                    className="Clear-Button"
                    disabled={true}
                    onClick={[Function]}
                  >
                    <button
                      className="Button"
                      disabled={true}
                      onClick={[Function]}
                    >
                      - Clear
                    </button>
                  </Button>
                </header>
                <ul
                  className="List"
                />
              </div>
            </FileList>
            <Button
              cancel={false}
              className="Bounty-Create-Upload"
              disabled={true}
              onClick={[Function]}
            >
              <button
                className="Button"
                disabled={true}
                onClick={[Function]}
              >
                Create Bounty
              </button>
            </Button>
          </div>
        </div>
      </BountyCreate>
    </div>
  </div>
</App>
`;

exports[`shows BountyInfo when at least one bounty & active selects it 1`] = `
<App>
  <div
    className="App"
  >
    <Sidebar
      active={0}
      bounties={
        Array [
          Object {
            "guid": "asdf",
          },
        ]
      }
      remove={[Function]}
      select={[Function]}
    >
      <div
        className="Sidebar"
      >
        <header
          className="Sidebar-Header"
        >
          <img
            alt="Polyswarm logo"
            src="/img/polyswarm.svg"
          />
        </header>
        <ul
          className="Sidebar-Content"
        >
          <ListItem
            active={true}
            className="item-0"
            key="asdf"
            onClick={[Function]}
            remove={[Function]}
          >
            <li
              className="ListItem active"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="ListItem-Child"
              >
                asdf
              </div>
            </li>
          </ListItem>
        </ul>
      </div>
    </Sidebar>
    <Header
      onClick={[Function]}
      title="asdf"
    >
      <header
        className="Header"
      >
        <h1>
          asdf
        </h1>
        <Button
          className="Header-Button"
          onClick={[Function]}
        >
          <button
            className="Button"
            onClick={[Function]}
          >
            + Bounty
          </button>
        </Button>
      </header>
    </Header>
    <div
      className="App-Content"
    >
      <BountyInfo
        bounty={
          Object {
            "guid": "asdf",
          }
        }
      >
        <div
          className="Bounty-Info"
        >
          <div
            className="Bounty-Info-Container"
          >
            <FileList
              className="Bounty-Info-Files"
              files={Array []}
              readonly={true}
            >
              <div
                className="File-List"
              >
                <header
                  className="File-List-Header"
                >
                  <p
                    className="File-List-Title"
                  >
                    0
                     
                    Files
                  </p>
                </header>
                <ul
                  className="List"
                />
              </div>
            </FileList>
          </div>
        </div>
      </BountyInfo>
    </div>
  </div>
</App>
`;

exports[`shows create when no bounties found 1`] = `null`;
